### DISCUSSION TITLE
FastAPI Query Parameters with datetime.date not updating default value?

### DISCUSSION THREAD

**Discussion by Marc-Girard** [fastapi/fastapi#9802] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
# Get a list of events from the logger DB
@app.get('/getlog')
async def get_log_from_database(
    bot_name: str,
    start_date: datetime.date = datetime.date.today(),
    end_date: datetime.date = datetime.date.today()
):
```

### Description

Hi everyone,

Just started using FastAPI and it is all been great but I'm struggling with an issue/bug that I can't seem to find a solution for. 

Now, I need the user to enter start/end dates, I want them to default to today's date if nothing is entered/provided in either one of the two.    

I seems like it is working but I just noticed that the default date provided by FastAPI is stuck in time.    It is always the date/day which we pushed the new image on our OpenShift host and never changes until we restart/update the image.   

Now, what can I do to refresh the default values with the correct "datetime.date.today()" value?  I did verify in my image itself via the console, the date/time are correct.

Any help is appreciated!

I included a screenshot.   

![Screenshot 2023-07-04 150239](https://github.com/tiangolo/fastapi/assets/53351429/bf89ea29-7d4a-4df4-b7d5-393099c01b35)

### Operating System

Linux

### Operating System Details

Running on Ubuntu 22.04 in a Docker image

### FastAPI Version

0.99.1

### Python Version

Python 3.11.4

**Comment by ahmedabdou14** [comment, marked as answer]
Actually the `openapi` specification file is created under `/openapi.json` path which is used to generate the html for the `swagger` documentation only once as the project builds. AFAIK, there is no way to dynamically change the default value except if you have a cron job that will play with fastapi internals in production and regenerate the swagger ui every _x_ amount of time.
(which is not a good idea 😅)

**Reply by Marc-Girard** [reply]
Thank you for your quick reply.   I simply changed the code to this and it works perfect now:

    async def get_log_from_database(bot_name: str, start_date: datetime.date, end_date: datetime.date | None = None):
    # No End Date provided?  Set it to today
    if end_date is None:
        end_date = datetime.date.today()

**Comment by FISHMANPET** [comment]
You've sort of worked around it, but I believe that in your original example, datetime.date.today() will run when the model is loaded (so basically, when your app starts up) rather than when it's accessed.  If you wanted to continue on that path, you'd want to use a pydantic Field and use `default_factory`.  There's an [example on the pydantic](https://docs.pydantic.dev/latest/usage/models/#fields-with-dynamic-default-values) site that's using time specifically.  I'm not sure what that will show in the docs, but it would at least work to actually set the default to the date when the model is created.

**Reply by ahmedabdou14** [reply]
Oh i thought he wanted the default value in the documentation to be current date.
if he used pydantic `Field` it would force him to have the dates inside the request body and not the query param

**Reply by FISHMANPET** [reply]
You know what, now that I've read that again, I realize the example code was an endpoint definition, not a pydantic model.  Whoops!  Though I think the same general comment applies, when you set a default value to the result of a function execution, that function will execute when the app starts, not when the function is called.  And since there's no default_factory parameter for the fastapi `Query()`, I think the best option is what Marc responded with yesterday.

**Reply by FISHMANPET** [reply]
And I see that just two days ago, Tiangolo added some more of the Field parameters to Query, Path, etc, including default_factory: https://github.com/tiangolo/fastapi/pull/9797
Doesn't look like that's released yet, and that will be part of the pydantic v2 support.  So pretty soon that might be an option!  You'd define the value something like this:

```
end_date: Annotated[datetime.date, Path(default_factory=datetime.date.today)]
```

**Reply by ahmedabdou14** [reply]
Oh wow, it has been merged already into the latest fastapi version. Such a good feature!, thank you for letting us know

**Comment by ahmedabdou14** [comment]
> And I see that just two days ago, Tiangolo added some more of the Field parameters to Query, Path, etc, including default_factory: #9797 Doesn't look like that's released yet, and that will be part of the pydantic v2 support. So pretty soon that might be an option! You'd define the value something like this:
> 
> ```
> end_date: Annotated[datetime.date, Path(default_factory=datetime.date.today)]
> ```

I migrated to pydantic v2 and fastapi 0.100.0 and tried this solution but unfortunately it is not working as anticipated.
This is what it looks like with `Query` & same result for `Path`

```python
@router.get("/")
async def test(
    date: Annotated[dt.datetime, Query(default_factory=dt.datetime.now)]
):
    return {"message": "Hello World!"}
```

![image](https://github.com/tiangolo/fastapi/assets/104530599/fe0e8769-6a8d-406f-ab8f-81a5c62aea5e)

However using `Body(default_factory=dt.datetime.now)` works as anticipated.

@tiangolo @Kludex Is this behavior of `Query` and `Param` `default_factory` normal?

**Reply by Kludex** [reply]
Please add the imports, and I'll check.

**Reply by FISHMANPET** [reply]
I tried this with fastapi 0.100.0 and pydantic 2.0.2
```python
from fastapi import FastAPI, APIRouter, Query, Body
import datetime as dt
from typing import Annotated

app = FastAPI()
router = APIRouter()

@router.get("/q")
async def test_q(
    date: Annotated[dt.datetime, Query(default_factory=dt.datetime.now)]
):
    return {"message": date}

@router.post("/b")
async def test_b(
    date: Annotated[dt.datetime, Body(default_factory=dt.datetime.now)]
):
    return {"message": date}

app.include_router(router)

```

Both endpoints worked as expected - when no query or no body was passed in, the returned value was the time right now - running it again and again returned different times.  In OpenAPI, the query looks as it does in @Ahmed-Abdou14's screenshot, there is no suggested value.  The body is a little different, in that the example value is the time when you load the OpenAPI docs.  I'm guessing the expectation is that the time when you open the OpenAPI doc is loaded as the suggestion for the query, as it is for the body.  However the fact that it doesn't looks like it would be a swagger-ui issue, not a fastapi issue.  In openapi.json, the "schema" for both the query and body are identical:
```json
"schema": {
"type": "string",
"format": "date-time",
"title": "Date"
}
```

**Reply by ahmedabdou14** [reply]
Thank you @FISHMANPET for the detailed response, small edit to your statement:
 `The body is a little different, in that the example value is the time when you load the OpenAPI docs`
Actually for the body parameter, the docs are updated to current time everytime i access the docs without reloading FastApi
